### PR TITLE
Generate: left-padding test, revisited

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Now that we understand why left-padding doesn't produce the exact same outputs (see [comment](https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535)), this PR revisits the left-padding test.

In a nutshell:
- no more retry loop, instead sets a proper error tolerance for the test
- adds an appropriate skip reason (when needed)
- removes many model-specific skips -- these models either pass the updated test or fall under a skip rule

### CODE REVIEW

**Comment by gante** [comment]
cc @zucchini-nlp -- this is a good example of us adding ugly flakiness-related loops because we didn't understand the cause for flakiness. Now that we understand that left-padding == not exactly the same results, we can write a clean test.

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -1834,49 +1834,68 @@ def test_generate_with_head_masking(self):
...
-                    no_failures = False
-                    break
-
-            self.assertTrue(no_failures)
+            # Without padding
+            model_kwargs = _prepare_model_kwargs(input_ids, attention_mask, signature)
+            next_logits_wo_padding = model(**model_kwargs).logits[:, -1, :]
+
+            # With left-padding (length 32)
+            pad_size = (input_ids.shape[0], 32)
```
Comment (gante) [inline tests/generation/test_utils.py:1890]: Note that the test is now much harder (padding size went from 1 to 32)

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -1834,49 +1834,68 @@ def test_generate_with_head_masking(self):
...
-        # 2 - To reduce the chance of false positives (i.e. passing when it should be failing), we run the check
-        # multiple times with random inputs, and it has to pass with all of them.
-        # NOTE: because of 2), there is some chance of false positives in this test.
+        # NOTE: left-padding results in small numerical differences. This is expected.
+        # See https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535
 
+        # First, filter out models that don't support left padding
+        # - The model must have generative capabilities
+        if len(self.all_generative_model_classes) == 0:
+            self.skipTest(reason="No generative model available.")
```
Comment (ArthurZucker) [inline tests/generation/test_utils.py:1843]:
```suggestion
            self.skipTest(reason="No generative architecture available for this model.")
```

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -1834,49 +1834,68 @@ def test_generate_with_head_masking(self):
...
+        decoder_only_classes = []
         for model_class in self.all_generative_model_classes:
             config, _, _, _ = self._get_input_ids_and_config()
             if config.is_encoder_decoder:
-                continue  # skip for encoder-decoder models -- they don't need left-padding compatibility
+                continue
+            else:
+                decoder_only_classes.append(model_class)
+        if len(decoder_only_classes) == 0:
+            self.skipTest(reason="No decoder-only model available.")
```
Comment (ArthurZucker) [inline tests/generation/test_utils.py:1854]:
```suggestion
            self.skipTest(reason="No decoder-only architecture available for this model.")
```

File: tests/generation/test_utils.py (hunk 1 above)
Comment (ArthurZucker) [reply tests/generation/test_utils.py:1890]: 💘

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -1834,49 +1834,68 @@ def test_generate_with_head_masking(self):
...
+            # With left-padding (length 32)
+            pad_size = (input_ids.shape[0], 32)
+            padding = torch.ones(pad_size, dtype=input_ids.dtype, device=torch_device) * config.pad_token_id
+            padded_input_ids = torch.cat((padding, input_ids), dim=1)
+            padded_attention_mask = torch.cat((torch.zeros_like(padding), attention_mask), dim=1)
+            model_kwargs = _prepare_model_kwargs(padded_input_ids, padded_attention_mask, signature)
+            next_logits_with_padding = model(**model_kwargs).logits[:, -1, :]
+
+            # They should result in very similar logits
+            self.assertTrue(torch.allclose(next_logits_wo_padding, next_logits_with_padding, atol=1e-5))
```
Comment (ArthurZucker) [inline tests/generation/test_utils.py:1898]: surprised that this is enough to make it pass!

**Review by ArthurZucker — APPROVED** [review]
Nice refactor!
